vshard/storage/init.lua

```diff
@@ -251,7 +251,17 @@ end
 -- exceptions are not allowed.
 --
 local function local_call(func_name, args)
-    return pcall(netbox_self_call, netbox_self, func_name, args)
+    -- If function was created with box.schema.func we call it via
+    -- func.call API, as net_box.self.call doesn't work with C stored
+    -- procedures up to Tarantool 3.0. If box.func is inaccessible,
+    -- which is true for all versions below 2.2, we fallback to
+    -- net_box.self.call too.
```

Comment:
If something is working one way for all versions < N and in another way for all versions > N, then this doesn't have to be a runtime check. Please, make `local_call()` depend on the core version. For earlier versions we can try the fallback approach. For newer ones - simply always use `netbox_self_call()` like before.

As "make it depend on core version" I mean that you can define `local_call()` differently depending on the version. Similar to how `util.index_min()` is defined.

Reply:
> If something is working one way for all versions < N and in another way for all versions > N, then this doesn't have to be a runtime check.

We've already done version dependent thing via runtime checks: [netbox_is_conn_dead](https://github.com/tarantool/vshard/blob/9fda3546ded3f74fa03dda371a984986e2b367d1/vshard/replicaset.lua#L165C1-L185).

We cannot use `netbox_self_call` for newer versions, as there's no version, where it works now. I don't know, when I'll do it ([ticket](https://github.com/tarantool/tarantool/issues/9131)). So, let's use runtime check for now.

Reply:
The only version dependent code we can make now: before 2.2.0 always use `netbox_self_call`, as there're no `box.func` API, in all other cases try to fallback. I don't really see much sense in such approach, but this can be done, if you insist

Reply:
> netbox_is_conn_dead

Here it is justified, because it has the fast-path which is version-agnostic and works almost always:
```Lua
    if conn.state == 'active' then
        return false
    end
```
In your case you've added branching on the path which is executed always, so this is different.

> We cannot use netbox_self_call for newer versions, as there's no version, where it works now

If this is not fixed even in master, then we can't merge nor test it until it is done in master. If this is already done in master, then you can always check even a not yet released version by the commit number. For example, the latest master I see is `3.0.0-beta1-5-ge844cd09b`, which means 5 commits after `3.0.0-beta1`. `vshard.version` module supports commit numbers, so you can do "if version >= 3.0.0-beta1-5 then use the new way, otherwise use the fallback approach".

Reply:
No, it's not done in master, the bug is still there (it's not related to release). Nonetheless, I don't get, why we're blocking the PR, if we can fix it in vshard right now, it's even tested.

So, at the end of a day we mark https://github.com/tarantool/vshard/issues/436 as blocked by https://github.com/tarantool/tarantool/issues/9131?

Reply:
Ok, I see the full picture now, thanks for the explanation. Lets finalize this PR and file a ticket to make the `local_call` depend on the version in the future.

Reply:
https://github.com/tarantool/tarantool/issues/9131 was fixed by the Core team. Let's make version banching